x-pack/plugins/data_visualizer/public/application/index_data_visualizer/search_strategy/requests/get_document_stats.ts

```diff
@@ -12,20 +12,98 @@ import { DataPublicPluginStart, ISearchOptions } from '@kbn/data-plugin/public';
 import seedrandom from 'seedrandom';
 import { isDefined } from '@kbn/ml-is-defined';
 import { buildBaseFilterCriteria } from '@kbn/ml-query-utils';
+import { AggregateQuery } from '@kbn/es-query';
+import { type UseCancellableSearch } from '@kbn/ml-cancellable-search';
 import { RANDOM_SAMPLER_PROBABILITIES } from '../../constants/random_sampler';
 import type {
   DocumentCountStats,
   OverallStatsSearchStrategyParams,
 } from '../../../../../common/types/field_stats';
+import { getSafeESQLName, isESQLQuery } from './esql_utils';
 
 const MINIMUM_RANDOM_SAMPLER_DOC_COUNT = 100000;
 const DEFAULT_INITIAL_RANDOM_SAMPLER_PROBABILITY = 0.000001;
 
+export const getESQLDocumentCountStats = async (
+  runRequest: UseCancellableSearch['runRequest'],
+  query: AggregateQuery,
+  filter?: estypes.QueryDslQueryContainer,
+  timeFieldName?: string,
+  intervalMs?: number,
+  searchOptions?: ISearchOptions
+): Promise<{ documentCountStats?: DocumentCountStats; totalCount: number }> => {
+  if (!isESQLQuery(query)) {
+    throw Error(
+      i18n.translate('xpack.dataVisualizer.esql.noQueryProvided', {
+        defaultMessage: 'No ES|QL query provided',
+      })
+    );
+  }
+  const esqlBaseQuery = query.esql;
+  let earliestMs = Infinity;
+  let latestMs = -Infinity;
+
+  if (timeFieldName) {
+    const aggQuery = ` | EVAL _timestamp_= TO_DOUBLE(DATE_TRUNC(${intervalMs} millisecond, ${getSafeESQLName(
+      timeFieldName
+    )}))
+    | stats rows = count(*) by _timestamp_
+    | LIMIT 10000`;
```

Comment:
Did you look at setting a default `LIMIT` if one is not supplied? I am finding the queries often time out when my query returns around 1M docs or more?

<img width="255" alt="image" src="https://github.com/elastic/kibana/assets/7405507/6d591842-8c4d-4f6a-b6dd-f9c4dd93f934">

Reply:
The 'limit' control makes a big difference to usability! One question - I noticed the doc count chart updates when I change the limit - is that necessary? Does the limit setting affect the doc count chart as well as the table?